plugin/core/windows.py

```diff
@@ -343,144 +90,177 @@ def __init__(
+                    if config.name == session.config.name and session.handles_path(file_name):
+                        handled = True
+                        break
+                if not handled:
+                    return config
+        else:
+            for config in configs:
+                handled = False
+                for session in self._sessions:
+                    if config.name == session.config.name:
+                        handled = True
+                        break
+                if not handled:
+                    return config
+        return None
 
-        def needed_configs(configs: Iterable[ClientConfig]) -> List[ClientConfig]:
-            new_configs = []
-            for c in configs:
-                if c.name not in self._sessions:
-                    new_configs.append(c)
-                else:
-                    session = next((s for s in self._sessions[c.name] if s.handles_path(file_path)), None)
-                    if session:
-                        if session.state != ClientStates.READY:
-                            debug('scheduling for delayed open, session {} not ready: {}'.format(c.name, file_path))
-                            self._open_after_initialize(view)
-                        else:
-                            debug('found ready session {} for {}'.format(c.name, file_path))
-                    else:
-                        debug('path not in existing {} session: {}'.format(c.name, file_path))
-                        new_configs.append(c)
-
-            return new_configs
-
-        # have all sessions for this document been started?
-        with self._initialization_lock:
-            # ViewLike vs sublime.View
-            startable_configs = needed_configs(self._configs.match_view(view))  # type: ignore
-            for config in startable_configs:
-                debug("window {} requests {} for {}".format(self._window.id(), config.name, file_path))
-                self.start(config, view)  # type: ignore
-
-    def start(self, config: ClientConfig, initiating_view: sublime.View) -> None:
+    def start_async(self, config: ClientConfig, initiating_view: sublime.View) -> None:
         file_path = initiating_view.file_name() or ''
         if not self._can_start_config(config.name, file_path):
-            debug('Already starting on this window:', config.name)
+            # debug('Already starting on this window:', config.name)
             return
         try:
             workspace_folders = sorted_workspace_folders(self._workspace.folders, file_path)
             plugin_class = get_plugin(config.name)
             if plugin_class is not None:
                 if plugin_class.needs_update_or_installation():
-                    self._window.status_message('Installing {} ...'.format(config.name))
+                    config.set_view_status(initiating_view, "installing...")
                     plugin_class.install_or_update()
```

Comment:
It would require more bookkeeping for maintaining the "starting session" (the `self._new_session`). The benefit of running this in the async thread is that less bookkeeping is needed. Because all pending listeners are now flowing through that thread. No locks needed. I don't see much benefit in spawning an extra thread. Perhaps we should also spawn an extra thread for starting a process? That also takes some time. And a thread for stopping a process. I'm not really in the business of "helping" other plugins in case they depend on async callbacks being invoked immediately.

Reply:
Note that with `asyncio` this all becomes very easy.

Reply:
I'm just concerned that installation will block other plugins and all STs *_async events from triggering. It just doesn't seem like a "fair play" to do that :)